// @aws-sdk/client-personalize-events/dist/cjs/models/models_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PutUsersRequest = exports.User = exports.ResourceNotFoundException = exports.PutItemsRequest = exports.Item = exports.PutEventsRequest = exports.Event = exports.InvalidInputException = void 0;
var InvalidInputException;
(function (InvalidInputException) {
    InvalidInputException.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InvalidInputException = exports.InvalidInputException || (exports.InvalidInputException = {}));
var Event;
(function (Event) {
    Event.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(Event = exports.Event || (exports.Event = {}));
var PutEventsRequest;
(function (PutEventsRequest) {
    PutEventsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PutEventsRequest = exports.PutEventsRequest || (exports.PutEventsRequest = {}));
var Item;
(function (Item) {
    Item.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(Item = exports.Item || (exports.Item = {}));
var PutItemsRequest;
(function (PutItemsRequest) {
    PutItemsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PutItemsRequest = exports.PutItemsRequest || (exports.PutItemsRequest = {}));
var ResourceNotFoundException;
(function (ResourceNotFoundException) {
    ResourceNotFoundException.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResourceNotFoundException = exports.ResourceNotFoundException || (exports.ResourceNotFoundException = {}));
var User;
(function (User) {
    User.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(User = exports.User || (exports.User = {}));
var PutUsersRequest;
(function (PutUsersRequest) {
    PutUsersRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PutUsersRequest = exports.PutUsersRequest || (exports.PutUsersRequest = {}));
